core/src/main/java/io/grpc/util/OutlierDetectionLoadBalancer.java

```diff
@@ -0,0 +1,1001 @@
+      this.config = config;
+    }
+
+    @Override
+    public void ejectOutliers(AddressTrackerMap trackerMap, long ejectionTimeMillis) {
+
+      // If we don't have the minimum amount of addresses the config calls for, then return.
+      if (trackerMap.size() < config.failurePercentageEjection.minimumHosts) {
+        return;
+      }
+
+      // If this address does not have enough volume to be considered, skip to the next one.
+      for (AddressTracker tracker : trackerMap.values()) {
+        // If an ejection now would take us past the max configured ejection percentagem stop here.
+        if (trackerMap.nextEjectionPercentage() > config.maxEjectionPercent) {
+          return;
+        }
+
+        if (tracker.inactiveVolume() < config.failurePercentageEjection.requestVolume) {
+          continue;
+        }
+
+        // If the failure rate is above the threshold, we should eject...
+        double maxFailureRate = ((double)config.failurePercentageEjection.threshold) / 100;
+        if (tracker.failureRate() > maxFailureRate) {
+          // ...but only enforce this based on the enforcement percentage.
+          if (new Random().nextInt(100) < config.failurePercentageEjection.enforcementPercentage) {
+            tracker.ejectSubchannels(ejectionTimeMillis);
+          }
+        }
+      }
+    }
+  }
+
+
+  /**
+   * The configuration for {@link OutlierDetectionLoadBalancer}.
+   */
+  public static final class OutlierDetectionLoadBalancerConfig {
+
+    final Long intervalSecs;
+    final Long baseEjectionTimeSecs;
+    final Long maxEjectionTimeSecs;
+    final Integer maxEjectionPercent;
+    final SuccessRateEjection successRateEjection;
+    final FailurePercentageEjection failurePercentageEjection;
+    final PolicySelection childPolicy;
+
+    private OutlierDetectionLoadBalancerConfig(Long intervalSecs,
+        Long baseEjectionTimeSecs,
+        Long maxEjectionTimeSecs, Integer maxEjectionPercent,
+        SuccessRateEjection successRateEjection,
+        FailurePercentageEjection failurePercentageEjection,
+        PolicySelection childPolicy) {
+      this.intervalSecs = intervalSecs;
+      this.baseEjectionTimeSecs = baseEjectionTimeSecs;
+      this.maxEjectionTimeSecs = maxEjectionTimeSecs;
+      this.maxEjectionPercent = maxEjectionPercent;
+      this.successRateEjection = successRateEjection;
+      this.failurePercentageEjection = failurePercentageEjection;
+      this.childPolicy = childPolicy;
+    }
+
+    /** Builds a new {@link OutlierDetectionLoadBalancerConfig}. */
+    public static class Builder {
+      Long intervalSecs = 10L;
+      Long baseEjectionTimeSecs = 30L;
+      Long maxEjectionTimeSecs = 30L;
+      Integer maxEjectionPercent = 10;
+      SuccessRateEjection successRateEjection;
+      FailurePercentageEjection failurePercentageEjection;
+      PolicySelection childPolicy;
+
+      /** The interval between outlier detection sweeps. */
```

Comment:
All of the setters should check the range of values, such as > 0 and where appropriate <= 100